File: backend/app/movement/collision_grid.py
"""
Collision Grid System

Represents the walkability of the map as a 2D boolean grid derived from:
- Terrain tile properties (walls, pits, rubble, etc.)
- Entity positions and blocking properties

Cached for performance; invalidated when entities move/spawn/despawn.
"""

from typing import Dict, List, Tuple, Optional, Set
from dataclasses import dataclass


@dataclass
class NavNode:
    """Represents a position on the navigation grid."""
    x: int
    y: int

    def __eq__(self, other):
        if not isinstance(other, NavNode):
            return False
        return self.x == other.x and self.y == other.y

    def __hash__(self):
        return hash((self.x, self.y))


class CollisionGrid:
    """
    2D grid representing walkability at each tile position.
    
    Structure:
    - walkable[y][x] = True if tile is traversable, False if blocked
    - Includes terrain blocking + entity occupancy
    """

    def __init__(self, width: int, height: int):
        self.width = width
        self.height = height
        # Initialize all tiles as walkable
        self.walkable: List[List[bool]] = [
            [True for _ in range(width)] for _ in range(height)
        ]
        self.version: int = 0  # Cache-busting version number

    def build_from_map(self, map_data) -> None:
        """
        Build collision grid from map data.
        
        Marks tiles as non-walkable if:
        - tile.blocks_movement == True
        - tile is a wall, pit, pillar, rubble
        - tile is a door with state="closed"
        
        Args:
            map_data: Map object with tiles dict and width/height
        """
        # Reset to all walkable
        self.walkable = [
            [True for _ in range(self.width)] for _ in range(self.height)
        ]

        # Iterate through all tiles and mark blocking tiles
        for (x, y), tile in map_data.tiles.items():
            if not self._is_tile_walkable(tile):
                if 0 <= x < self.width and 0 <= y < self.height:
                    self.walkable[y][x] = False

        self.version += 1

    def update_entity_blocking(self, entities: List) -> None:
        """
        Update collision grid with dynamic entity blocking.
        
        Entities with blocks_movement=True occupy their tile and make it unwalkable.
        This is called after entity positions change.
        
        Args:
            entities: List of entity objects with x, y, blocks_movement properties
        """
        # First, rebuild terrain blocking (don't rely on cumulative state)
        # This is a simplified version; in production, you'd rebuild from map_data
        # For now, we track entity blocking separately
        
        for entity in entities:
            if hasattr(entity, "blocks_movement") and entity.blocks_movement:
                x = getattr(entity, "x", None)
                y = getattr(entity, "y", None)
                if x is not None and y is not None:
                    if 0 <= x < self.width and 0 <= y < self.height:
                        self.walkable[y][x] = False

        self.version += 1

    def is_walkable(self, x: int, y: int) -> bool:
        """Check if a tile is walkable."""
        if not (0 <= x < self.width and 0 <= y < self.height):
            return False
        return self.walkable[y][x]

    def get_neighbors(
        self, x: int, y: int, include_diagonal: bool = True
    ) -> List[NavNode]:
        """
        Get walkable neighbors of a tile.
        
        Enforces diagonal corner blocking:
        - A diagonal move requires BOTH adjacent orthogonal tiles to be walkable.
        - E.g., moving NE from (x, y) to (x+1, y-1) requires:
          - (x+1, y) walkable (right)
          - (x, y-1) walkable (up)
        
        Args:
            x, y: Tile coordinates
            include_diagonal: Whether to include diagonal neighbors
            
        Returns:
            List of walkable NavNode neighbors
        """
        neighbors: List[NavNode] = []

        # Orthogonal neighbors (always check these)
        # Right
        if self.is_walkable(x + 1, y):
            neighbors.append(NavNode(x + 1, y))
        # Left
        if self.is_walkable(x - 1, y):
            neighbors.append(NavNode(x - 1, y))
        # Down
        if self.is_walkable(x, y + 1):
            neighbors.append(NavNode(x, y + 1))
        # Up
        if self.is_walkable(x, y - 1):
            neighbors.append(NavNode(x, y - 1))

        if include_diagonal:
            # Diagonal neighbors with corner blocking rules
            # NE: (x+1, y-1) — requires (x+1, y) AND (x, y-1) walkable
            if self.is_walkable(x + 1, y - 1) and self.is_walkable(x + 1, y) and self.is_walkable(x, y - 1):
                neighbors.append(NavNode(x + 1, y - 1))
            # NW: (x-1, y-1) — requires (x-1, y) AND (x, y-1) walkable
            if self.is_walkable(x - 1, y - 1) and self.is_walkable(x - 1, y) and self.is_walkable(x, y - 1):
                neighbors.append(NavNode(x - 1, y - 1))
            # SE: (x+1, y+1) — requires (x+1, y) AND (x, y+1) walkable
            if self.is_walkable(x + 1, y + 1) and self.is_walkable(x + 1, y) and self.is_walkable(x, y + 1):
                neighbors.append(NavNode(x + 1, y + 1))
            # SW: (x-1, y+1) — requires (x-1, y) AND (x, y+1) walkable
            if self.is_walkable(x - 1, y + 1) and self.is_walkable(x - 1, y) and self.is_walkable(x, y + 1):
                neighbors.append(NavNode(x - 1, y + 1))

        return neighbors

    @staticmethod
    def _is_tile_walkable(tile) -> bool:
        """
        Determine if a tile is walkable based on its type and state.
        
        Returns True if walkable, False if blocking.
        """
        # Check explicit blocks_movement property first
        if hasattr(tile, "blocks_movement"):
            blocks = getattr(tile, "blocks_movement", False)
            # Special case: doors block only if state == "closed"
            tile_type = getattr(tile, "tile_type", getattr(tile, "type", "floor"))
            if tile_type == "door":
                state = getattr(tile, "state", None)
                return state != "closed"
            return not blocks

        # Fallback to tile type (check both tile_type and type for compatibility)
        tile_type = getattr(tile, "tile_type", getattr(tile, "type", "floor"))
        blocking_types = {"wall", "pit", "pillar", "rubble"}
        
        if tile_type in blocking_types:
            return False
        
        # Doors block if closed
        if tile_type == "door":
            state = getattr(tile, "state", None)
            return state != "closed"

        return True

File: frontend/src/lib/systems/movement/collisionGrid.ts
/**
 * Collision Grid System (Frontend)
 * 
 * Client-side collision grid for fast local pathfinding validation.
 * Mirrors backend structure for consistency.
 */

export interface NavNode {
  x: number;
  y: number;
}

export class CollisionGrid {
  width: number;
  height: number;
  walkable: boolean[][];
  movementCost: number[][];
  version: number = 0;

  constructor(width: number, height: number) {
    this.width = width;
    this.height = height;
    // Initialize all tiles as walkable
    this.walkable = Array(height)
      .fill(null)
      .map(() => Array(width).fill(true));
    this.movementCost = Array(height)
      .fill(null)
      .map(() => Array(width).fill(1));
  }

  /**
   * Build collision grid from map tile data.
   * Gameplay scenes are traversal-grid authoritative.
   */
  buildFromMap(
    _tiles: any[],
    width: number,
    height: number,
    traversalGrid?: any | null,
    options?: { mapMode?: string | null },
  ): void {
    // Reset to all walkable
    this.walkable = Array(height)
      .fill(null)
      .map(() => Array(width).fill(true));
    this.movementCost = Array(height)
      .fill(null)
      .map(() => Array(width).fill(1));

    const mapMode = typeof options?.mapMode === 'string' ? options.mapMode.trim().toLowerCase() : ''
    const aiMode = mapMode === 'ai_generated_image'
    const hasTraversalGrid = Boolean(
      traversalGrid
      && Array.isArray(traversalGrid.cells)
      && traversalGrid.width_cells
      && traversalGrid.height_cells,
    )

    if (hasTraversalGrid) {
      const scaleX = Number(traversalGrid.width_cells) / Math.max(1, width)
      const scaleY = Number(traversalGrid.height_cells) / Math.max(1, height)
      const cellsByKey = new Map<string, any>()
      for (const cell of traversalGrid.cells) {
        cellsByKey.set(`${cell.x},${cell.y}`, cell)
      }

      for (let y = 0; y < height; y++) {
        for (let x = 0; x < width; x++) {
          const sx0 = Math.floor(x * scaleX)
          const sx1 = Math.max(sx0, Math.floor((x + 1) * scaleX) - 1)
          const sy0 = Math.floor(y * scaleY)
          const sy1 = Math.max(sy0, Math.floor((y + 1) * scaleY) - 1)
          let anyTraversable = false
          let maxCost = 1

          for (let sy = sy0; sy <= sy1; sy++) {
            for (let sx = sx0; sx <= sx1; sx++) {
              const cell = cellsByKey.get(`${sx},${sy}`)
              if (!cell) continue
              if (cell.traversable) {
                anyTraversable = true
                const candidate = Number.isFinite(cell.movement_cost) ? Number(cell.movement_cost) : 1
                maxCost = Math.max(maxCost, candidate)
              }
            }
          }

          this.walkable[y][x] = anyTraversable
          this.movementCost[y][x] = Math.max(1, maxCost)
        }
      }

      this.version++
      return
    }

    // Legacy tile fallback is disabled for gameplay scenes. Lock movement until traversal arrives.
    this.walkable = Array(height)
      .fill(null)
      .map(() => Array(width).fill(false));
    this.movementCost = Array(height)
      .fill(null)
      .map(() => Array(width).fill(Number.POSITIVE_INFINITY));

    if (aiMode) {
      console.warn('[CollisionGrid] AI map mode active but traversal_grid missing; movement grid locked until traversal data arrives.');
    } else {
      console.warn('[CollisionGrid] Non-AI map mode encountered in gameplay scene; legacy tile collision is disabled.');
    }

    this.version++;
  }

  /**
   * Update collision grid with dynamic entity blocking.
   * 
   * Entities with blocks_movement=True occupy their tile.
   */
  updateEntityBlocking(entities: any[]): void {
    for (const entity of entities) {
      if (entity.blocks_movement && entity.x >= 0 && entity.x < this.width && entity.y >= 0 && entity.y < this.height) {
        this.walkable[entity.y][entity.x] = false;
      }
    }

    this.version++;
  }

  /**
   * Check if a tile is walkable.
   */
  isWalkable(x: number, y: number): boolean {
    if (!(x >= 0 && x < this.width && y >= 0 && y < this.height)) {
      return false;
    }
    return this.walkable[y][x];
  }

  getMovementCost(x: number, y: number): number {
    if (!(x >= 0 && x < this.width && y >= 0 && y < this.height)) {
      return Number.POSITIVE_INFINITY
    }
    return this.movementCost[y][x]
  }

  /**
   * Get walkable neighbors of a tile.
   * 
   * Enforces diagonal corner blocking:
   * - Diagonal moves require BOTH adjacent orthogonal tiles to be walkable
   */
  getNeighbors(x: number, y: number, includeDiagonal: boolean = true): NavNode[] {
    const neighbors: NavNode[] = [];

    // Orthogonal neighbors
    if (this.isWalkable(x + 1, y)) neighbors.push({ x: x + 1, y });
    if (this.isWalkable(x - 1, y)) neighbors.push({ x: x - 1, y });
    if (this.isWalkable(x, y + 1)) neighbors.push({ x, y: y + 1 });
    if (this.isWalkable(x, y - 1)) neighbors.push({ x, y: y - 1 });

    if (includeDiagonal) {
      // NE: requires (x+1, y) AND (x, y-1)
      if (
        this.isWalkable(x + 1, y - 1) &&
        this.isWalkable(x + 1, y) &&
        this.isWalkable(x, y - 1)
      ) {
        neighbors.push({ x: x + 1, y: y - 1 });
      }
      // NW: requires (x-1, y) AND (x, y-1)
      if (
        this.isWalkable(x - 1, y - 1) &&
        this.isWalkable(x - 1, y) &&
        this.isWalkable(x, y - 1)
      ) {
        neighbors.push({ x: x - 1, y: y - 1 });
      }
      // SE: requires (x+1, y) AND (x, y+1)
      if (
        this.isWalkable(x + 1, y + 1) &&
        this.isWalkable(x + 1, y) &&
        this.isWalkable(x, y + 1)
      ) {
        neighbors.push({ x: x + 1, y: y + 1 });
      }
      // SW: requires (x-1, y) AND (x, y+1)
      if (
        this.isWalkable(x - 1, y + 1) &&
        this.isWalkable(x - 1, y) &&
        this.isWalkable(x, y + 1)
      ) {
        neighbors.push({ x: x - 1, y: y + 1 });
      }
    }

    return neighbors;
  }
}
